sale_order_lot_selection/model/sale.py

```diff
@@ -44,24 +44,19 @@ def product_id_change_with_wh(
             warehouse_id=warehouse_id)
 
         available_lots = []
-        lot_model = self.env['stock.production.lot']
-        product_model = self.env['product.product']
-        product_obj = product_model.browse(product)
         location = self.env['stock.warehouse'].browse(
             warehouse_id).lot_stock_id
-        lots = lot_model.search(
-            [('product_id', '=', product)])
-        for lot in lots:
-            # for the selected product, search for every associated lot
-            # for every lot, check if it is available (in location.id)
-            # if it is, add it to selectable lots
-            qty_res = product_obj.with_context({
-                'lot_id': lot.id,
-                'location': location.id,
-                })._product_available()
-            if qty_res[product]['qty_available'] > 0:
-                if lot.id not in available_lots:
-                    available_lots.append(lot.id)
+        locations = self.env['stock.location'].search([
+            ('id', 'child_of', location.id),
+            ])
+        # Search all lot existing lot for the product and location selected
+        self._cr.execute("""SELECT lot_id
```

Comment:
You can use quants for searching, but without SQL. The performance will be for sure still good. You can use read_group function if you want to increase it. This way, you don't overpass multi-company rules, and so on.